test/integration/apiserver/field_validation_test.go

```diff
@@ -2887,6 +2917,103 @@ func testFieldValidationApplyUpdateCRDSchemaless(t *testing.T, rest rest.Interfa
 	}
 }
 
+func testFinalizerValidationApplyCreateAndUpdateCRD(t *testing.T, rest rest.Interface, gvk schema.GroupVersionKind, gvr schema.GroupVersionResource) {
+	var testcases = []struct {
+		name                   string
+		finalizer              string
+		updated_finalizer      string
+		opts                   metav1.PatchOptions
+		strictDecodingError    string
+		strictDecodingWarnings []string
+	}{
+		{
+			name:      "create-crd-with-invalid-finalizer",
+			finalizer: "invalid-finalizer",
+			strictDecodingWarnings: []string{
+				`metadata: Invalid value: "%s": prefer a domain-qualified finalizer name to avoid accidental conflicts with other finalizer writers`,
+			},
+		},
+		{
+			name:      "create-crd-with-valid-finalizer",
+			finalizer: "kubernetes.io/valid-finalizer",
+		},
+		{
+			name:              "update-crd-with-invalid-finalizer",
+			finalizer:         "invalid-finalizer",
+			updated_finalizer: "another-invalid-finalizer",
+			strictDecodingWarnings: []string{
+				`metadata: Invalid value: "%s": prefer a domain-qualified finalizer name to avoid accidental conflicts with other finalizer writers`,
+			},
+		},
+		{
+			name:              "update-crd-with-valid-finalizer",
+			finalizer:         "kubernetes.io/valid-finalizer",
+			updated_finalizer: "kubernetes.io/another-valid-finalizer",
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			kind := gvk.Kind
+			apiVersion := gvk.Group + "/" + gvk.Version
+
+			// create the CR as specified by the test case
+			name := fmt.Sprintf("apply-create-crd-%s", tc.name)
+			applyCreateBody := []byte(fmt.Sprintf(crdApplyFinalizerBody, apiVersion, kind, name, tc.finalizer))
+
+			req := rest.Patch(types.ApplyPatchType).
+				AbsPath("/apis", gvr.Group, gvr.Version, gvr.Resource).
+				Name(name).
+				Param("fieldManager", "apply_test").
+				VersionedParams(&tc.opts, metav1.ParameterCodec)
+			result := req.Body(applyCreateBody).Do(context.TODO())
+			if result.Error() == nil && tc.strictDecodingError != "" {
+				t.Fatalf("received nil error when expecting: %q", tc.strictDecodingError)
+			}
+			if result.Error() != nil && (tc.strictDecodingError == "" || !strings.HasSuffix(result.Error().Error(), tc.strictDecodingError)) {
+				t.Fatalf("expected error: %q, got: %v", tc.strictDecodingError, result.Error())
+			}
+
+			if len(result.Warnings()) != len(tc.strictDecodingWarnings) {
+				for _, r := range result.Warnings() {
+					t.Logf("received warning: %v", r)
+				}
+				t.Fatalf("unexpected number of warnings, expected: %d, got: %d", len(tc.strictDecodingWarnings), len(result.Warnings()))
+			}
+			for i, strictWarn := range tc.strictDecodingWarnings {
+				if fmt.Sprintf(strictWarn, tc.finalizer) != result.Warnings()[i].Text {
+					t.Fatalf("expected warning: %s, got warning: %s", strictWarn, result.Warnings()[i].Text)
+				}
+			}
```

Comment:
if you inline the exact warning text you expect, you can simplify this to `if !reflect.DeepEqual(tc.expectCreateWarnings, result.Warnings()) { ... }`

Reply:
`tc.expectCreateWarnings`(`[]string` type) has different type with `result.Warnings()`(`[]net.WarningHeader` type)

Reply:
ah, ok